manager/orchestrator/taskreaper/task_reaper.go

```diff
@@ -148,6 +148,12 @@ func (tr *TaskReaper) Run(ctx context.Context) {
 				if t.DesiredState == api.TaskStateRemove && t.Status.State >= api.TaskStateCompleted {
 					tr.cleanup = append(tr.cleanup, t.ID)
 				}
+
+				// Also clean up tasks which have not yet been assigned but have been
+				// marked for shutdown (likley because of a service update).
+				if t.Status.State < api.TaskStateAssigned && t.DesiredState == api.TaskStateShutdown {
```

Comment:
Based on IRL discussion with @anshulpundir, we think there can be a race condition here (see also related discussion in https://github.com/docker/swarmkit/pull/2557):

I'm unable to come up with an exact scenario, but it's something to think about cc @dperny @cyli. If y'all think this looks good, we can merge it.

Reply:
UPDATE: Based on another IRL discussion with @anshulpundir, we concluded that this change looks fine, in conjunction with the change in `manager/scheduler/scheduler.go`. The main idea is that as long as current state is `<= PENDING`, the task will not have reached the agent. Hence, it is safe to more aggressively remove it in the task reaper. After this change, we think it is also safe to merge https://github.com/docker/swarmkit/pull/2557.